src/machine/usb_nrf52840.go

```diff
@@ -12,22 +12,100 @@ import (
 
 // USBCDC is the USB CDC aka serial over USB interface on the nRF52840
 type USBCDC struct {
-	Buffer       *RingBuffer
-	interrupt    interrupt.Interrupt
-	initcomplete bool
+	Buffer            *RingBuffer
+	interrupt         interrupt.Interrupt
+	initcomplete      bool
+	TxIdx             volatile.Register8
+	waitTxc           bool
+	waitTxcRetryCount uint8
+	sent              bool
+}
+
+const (
+	usbcdcTxSizeMask          uint8 = 0x3F
+	usbcdcTxBankMask          uint8 = ^usbcdcTxSizeMask
+	usbcdcTxBank1st           uint8 = 0x00
+	usbcdcTxBank2nd           uint8 = usbcdcTxSizeMask + 1
+	usbcdcTxMaxRetriesAllowed uint8 = 5
+)
+
+// Flush flushes buffered data.
+func (usbcdc *USBCDC) Flush() error {
+	if usbLineInfo.lineState > 0 {
+		idx := usbcdc.TxIdx.Get()
+		sz := idx & usbcdcTxSizeMask
+		bk := idx & usbcdcTxBankMask
+		if 0 < sz {
+
+			if usbcdc.waitTxc {
+				// waiting for the next flush(), because the transmission is not complete
+				usbcdc.waitTxcRetryCount++
+				return nil
+			}
+			usbcdc.waitTxc = true
+			usbcdc.waitTxcRetryCount = 0
+
+			// set the data
+			enterCriticalSection()
+			sendViaEPIn(
+				usb_CDC_ENDPOINT_IN,
+				&udd_ep_in_cache_buffer[usb_CDC_ENDPOINT_IN][bk],
+				int(sz),
+			)
+			if bk == usbcdcTxBank1st {
+				usbcdc.TxIdx.Set(usbcdcTxBank2nd)
+			} else {
+				usbcdc.TxIdx.Set(usbcdcTxBank1st)
+			}
+
+			UART0.sent = true
+		}
+	}
+	return nil
 }
 
 // WriteByte writes a byte of data to the USB CDC interface.
 func (usbcdc USBCDC) WriteByte(c byte) error {
 	// Supposedly to handle problem with Windows USB serial ports?
 	if usbLineInfo.lineState > 0 {
-		enterCriticalSection()
-		udd_ep_in_cache_buffer[usb_CDC_ENDPOINT_IN][0] = c
-		sendViaEPIn(
-			usb_CDC_ENDPOINT_IN,
-			&udd_ep_in_cache_buffer[usb_CDC_ENDPOINT_IN][0],
-			1,
-		)
+		ok := false
+		for {
+			mask := interrupt.Disable()
+
+			idx := UART0.TxIdx.Get()
```

Comment:
In this function, access is made in the form UART0.XXX instead of usbcdc.XXX.
I don't know why the usbcdc.XXX access method is slow, but it is.

https://twitter.com/sago35tk/status/1363850766486675458

Reply:
The reelboard test was failing, so I rewrote it to USB instead of UART0.